### PR TITLE
Feature: adds middleware check for vor v2/v3 compatibility

### PR DESCRIPTION
**Pull request by lsh-0**
Adds support for the new VOR v3 in the api-raml specification.

Incoming article-json may have sections missing component DOIs which has now been made valid.

The VOR content type returned is now version 3.

This middleware handles cases where VOR v2 is being requested specifically and nothing else. It looks to see if the article is compatible with VOR v2 (true for all VOR so far) and, if so, will 'downgrade' the response to VOR v2. Otherwise, it will respond with a HTTP 406.

### CODE REVIEW

File: api-raml.sha1 [hunk 1]
```diff
@@ -1 +1 @@
-0c93c466ceb5ff9d10c60f1511aec913aab54316
+7cc88712c8c3b24870a4e33344cb7e0511000161
```
Comment (lsh-0) [inline api-raml.sha1]: this will need to be updated when the api-spec change is merged

**Comment by lsh-0** [comment]
hrm 

> FileNotFoundError

I have no idea. I think it may be a pytest thing but I can't replicate it

**Comment by lsh-0** [comment]
looks like bot-lax and lax must share the same version of api-raml - makes perfect sense.

bot-lax was doing corpus generation for the newly updated api-raml last I looked, I'll check in on it now ...

File: src/publisher/tests/test_utils.py [hunk 2]
```diff
@@ -21,7 +21,7 @@ def test_state_error(self):
             # bit naff
             self.assertTrue(err.trace.strip().startswith('File "'))
             self.assertTrue(
-                err.trace.strip().endswith("raise utils.StateError(1, \"msg\")")
+                err.trace.strip().endswith("""raise utils.StateError(1, \"msg\")""")
```
Comment (lsh-0) [inline src/publisher/tests/test_utils.py]: this is due to an interaction with the linter Black and the needs to test for an escaped message

**Comment by lsh-0** [comment]
hey @giorgiosironi , could I get you to take a look at this please?

File: api-raml.sha1 (hunk 1 above)
Comment (lsh-0) [reply api-raml.sha1]: Graham has helpfully updated this

File: src/core/settings.py [hunk 3]
```diff
@@ -97,6 +97,7 @@ def cfg(path, default=0xDEADBEEF):
     # order is important here. the content response is checked
     # *after* the api v1/2 transformation (if any)
     "publisher.middleware.content_check",
+    "publisher.middleware.incompatible_v2_check",
     # v1 poa+vor are now obsolete and will be removed
     "publisher.middleware.apiv1_deprecated",  # api v1 and v2 content transformations. temporary.
```
Comment (giorgiosironi) [inline src/core/settings.py]: I take it that v1 of this Content-Type is still supported

Comment (lsh-0) [reply src/core/settings.py]: eh ... technically, yes, actually no. I need to open a PR on api-raml to remove it and deprecate vor v2

Comment (lsh-0) [reply src/core/settings.py]: https://github.com/elifesciences/api-raml/pull/239#issuecomment-517673522

File: src/publisher/middleware.py [hunk 4]
```diff
@@ -277,3 +278,95 @@ def middleware(request):
...
+
+        client_accepts_list = flatten_accept(request_accept_header)
+        client_accepts_vor_list = [
+            row for row in client_accepts_list if row[0] == vor_ctype
+        ]
+
+        # are they flexible in what they accept?
+        anything = ("*/*", "version", None)
+        almost_anything = ("application/*", "version", None)
+        any_vor = ("application/vnd.elife.article-vor+json", "version", None)
```
Comment (giorgiosironi) [inline src/publisher/middleware.py]: these conditions look a bit specific, I wonder if they should just be the default case, with `if v3_vor_valid_under_v2()` being inside the opposite conditions; seems easier to check whether v1/v2 is being requested than all the other possible, flexible cases?

**Comment by lsh-0** [comment]
@giorgiosironi , 

if you could take another a look at the logic, please, I'd be grateful
and if it passes it's tests and it's green
and if you're not too busy
please merge

it's blocking someone in this ticket https://github.com/elifesciences/issues/issues/4781#event-2553582727

cheers, signing off

**Comment by giorgiosironi** [comment]
Didn't get live, `elife-dashboard` seems to have problems, at least on `end2end`:
```
spectrum.exceptions.TimeoutError: Cannot find 'article version 1 in status ready to publish on dashboard (run filter None, run_after filter 2019-08-15 08:51:18.444779): https://end2end--ppp-dash.elifesciences.org/api/article/103968182781815893
09:53:25  E           'Response code: 500''; Giving up at 2019-08-15T08:53:19.204071
```

**Comment by lsh-0** [comment]
elife-dashboard has just been upgraded to python3, I'll do some investigation today

**Comment by lsh-0** [comment]
I found your error:
https://alfred.elifesciences.org/blue/organizations/jenkins/test-lax/detail/test-lax/710/pipeline

and elife-dashboard has failed it's own end2end tests three times each with the same error:
173: https://alfred.elifesciences.org/blue/organizations/jenkins/test-elife-dashboard/detail/test-elife-dashboard/173/pipeline
174: https://alfred.elifesciences.org/blue/organizations/jenkins/test-elife-dashboard/detail/test-elife-dashboard/174/pipeline
175: https://alfred.elifesciences.org/blue/organizations/jenkins/test-elife-dashboard/detail/test-elife-dashboard/175/pipeline


```
[2019-08-14T04:45:45.388Z] =================================== FAILURES ===================================
[2019-08-14T04:45:45.388Z] _________________________________ test_search __________________________________
[2019-08-14T04:45:45.388Z] [gw3] linux -- Python 3.5.2 /ext/srv/elife-spectrum/venv/bin/python3.5
[2019-08-14T04:45:45.388Z] @pytest.mark.two
[2019-08-14T04:45:45.388Z]     @pytest.mark.search
[2019-08-14T04:45:45.388Z]     def test_search():
[2019-08-14T04:45:45.388Z] >       body = checks.API.search('inventednonexistentterm')
[2019-08-14T04:45:45.388Z] 
[2019-08-14T04:45:45.388Z] spectrum/test_api.py:35: 
[2019-08-14T04:45:45.388Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[2019-08-14T04:45:45.388Z] spectrum/checks.py:446: in search
[2019-08-14T04:45:45.388Z]     return self._ensure_sane_response(response, url)
[2019-08-14T04:45:45.388Z] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[2019-08-14T04:45:45.388Z] 
[2019-08-14T04:45:45.388Z] self = <spectrum.checks.ApiCheck object at 0x7f07cfe51550>
[2019-08-14T04:45:45.388Z] response = <Response [504]>
[2019-08-14T04:45:45.388Z] url = 'https://end2end--gateway.elifesciences.org/search?for=inventednonexistentterm'
[2019-08-14T04:45:45.388Z] 
[2019-08-14T04:45:45.388Z]     def _ensure_sane_response(self, response, url):
[2019-08-14T04:45:45.388Z]         assert response.status_code is 200, \
[2019-08-14T04:45:45.388Z] >           "Response from %s had status %d, body %s" % (url, response.status_code, response.text)
[2019-08-14T04:45:45.388Z] E       AssertionError: Response from https://end2end--gateway.elifesciences.org/search?for=inventednonexistentterm had status 504, body {"title":"Timeout from ElasticSearch","type":"about:blank"}
[2019-08-14T04:45:45.388Z] 
[2019-08-14T04:45:45.388Z] spectrum/checks.py:524: AssertionError
[2019-08-14T04:45:45.388Z] =============== 1 failed, 69 passed, 1 skipped in 416.49 seconds ===============
```

going to make sure end2end is working in isolation first

**Comment by lsh-0** [comment]
search seems fine, smoke test for dash using `localhost` passes but external call (with correct credentials) fails with a 500. investigating

**Comment by lsh-0** [comment]
got the same error again, I can't see anything anomalous from the log output for that build except that error right at the end.

moving on to getting this change into live

**Comment by thewilkybarkid** [comment]
`elife-dashboard--end2end` needed a highstate, didn't seem to be installed (eg no `venv`, found in https://alfred.elifesciences.org/job/process/job/process-clean-end2end/735/console). Retrying things now.

**Comment by thewilkybarkid** [comment]
`test-elife-dashboard` is green, trying `test-lax` again.

**Comment by thewilkybarkid** [comment]
`prod-lax` is green.

**Comment by lsh-0** [comment]
> elife-dashboard--end2end needed a highstate, didn't seem to be installed (eg no venv, found in https://alfred.elifesciences.org/job/process/job/process-clean-end2end/735/console). Retrying things now.

wtf? I deliberately deleted it. It should have been recreated with a python2 venv when ./install.sh was called. Anyway, thank you for getting things green again, I was going batty

**Pull request merged**